streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -630,57 +631,185 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
   )(implicit trace: Trace): ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone] =
-    ZChannel.unwrapScopedWith { scope =>
-      for {
-        input       <- SingleProducerAsyncInput.make[InErr, InElem, InDone]
-        queueReader  = ZChannel.fromInput(input)
-        queue       <- Queue.bounded[ZIO[Env1, OutErr1, Either[OutDone, OutElem2]]](n)
-        _           <- scope.addFinalizer(queue.shutdown)
-        errorSignal <- Promise.make[OutErr1, Nothing]
-        permits     <- Semaphore.make(n.toLong)
-        pull        <- (queueReader >>> self).toPullIn(scope)
-        _ <- pull
-               .foldCauseZIO(
-                 cause => queue.offer(ZIO.refailCause(cause)),
-                 {
-                   case Left(outDone) =>
-                     permits.withPermits(n.toLong)(ZIO.unit).interruptible *> queue.offer(ZIO.succeed(Left(outDone)))
-                   case Right(outElem) =>
-                     for {
-                       p     <- Promise.make[OutErr1, OutElem2]
-                       latch <- Promise.make[Nothing, Unit]
-                       _     <- queue.offer(p.await.map(Right(_)))
-                       _ <- permits.withPermit {
-                              latch.succeed(()) *>
-                                ZIO.uninterruptibleMask { restore =>
-                                  restore(errorSignal.await) raceFirstAwait restore(f(outElem))
-                                }
-                                  .tapErrorCause(errorSignal.failCause)
-                                  .intoPromise(p)
-                            }.forkIn(scope)
-                       _ <- latch.await
-                     } yield ()
-                 }
-               )
-               .forever
-               .interruptible
-               .forkIn(scope)
-      } yield {
-        lazy val consumer: ZChannel[Env1, Any, Any, Any, OutErr1, OutElem2, OutDone] =
-          ZChannel.unwrap[Env1, Any, Any, Any, OutErr1, OutElem2, OutDone] {
-            queue.take.flatten.foldCause(
-              ZChannel.refailCause,
-              {
-                case Left(outDone)  => ZChannel.succeedNow(outDone)
-                case Right(outElem) => ZChannel.write(outElem) *> consumer
-              }
+    mapOutZIOPar[Env1, OutErr1, OutElem2](n, n)(f)
+
+  final def mapOutZIOPar[Env1 <: Env, OutErr1 >: OutErr, OutElem2](n: Int, bufferSize: Int)(
+    f: OutElem => ZIO[Env1, OutErr1, OutElem2]
+  )(implicit trace: Trace): ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone] = {
+    val z: ZIO[Any, Nothing, ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone]] = for {
+      input         <- SingleProducerAsyncInput.make[InErr, InElem, InDone]
+      queueReader    = ZChannel.fromInput(input)
+      queue         <- Queue.bounded[Fiber[Option[OutErr1], OutElem2]](bufferSize)
+      permits       <- zio.Semaphore.make(n)
+      failureSignal <- Promise.make[Option[OutErr1], Nothing]
```

Comment:
not sure what u mean by 'loosing errors', AFAIK this preserves the same behavior as the original implementation, hence in case of failures the stream fails with the first noticeable failure (in case of race, an arbitrary 'winner'). in fact the original impl also used the errorSignal promise both for interrupting 'surviving' fibers and for signaling downstream of the error (the effect in the queue reflects that).

re. changing the failure signal to `Promise[Nothing, Unit]`, this will fail some of the existing unit-tests. basically in case of parallelism with N>1, a case when the first effects is blocking/still running but another effect fails should result with the stream failing with the 'later' failure (with all pending fibers effectively interrupted). this means either the reader or the writers must race with the failure signal, doing this on the reader side is a bit simpler but resulted with a performance penalty (roughly speaking you need to race this per record, better do this in parallel. backed by benchmark numbers).

re. your concerns about the fibers failing with interrupt instead of the actual failure: two cases when fibers are interrupted:
1. the entire stream is interrupted, I think we can't expect another failure in this scenario
2. when another fiber fails and manages to complete the failure signal promise, since the executing fibers are all raced with the failure signal, relying on `raceWith` behavior, I explicitly interrupt the looser and propagate the error from the winning side, so I can't c an interruption winning over an actual failure in this scenario

Reply:
yes i can see errors can be lost in the current implementation too, so it's fine if you don't try to address that. you can see how multiple failing fibers are handled with `ZIO.forEachPar` here: https://scastie.scala-lang.org/HtsjbCtTSoCoT5UBuKw28Q and if you want to learn more about how this works: https://zio.dev/reference/core/cause#cause-internals

i wonder if the performance hit could be overcome by changing the one element at a time nature of your solution. even with your current implementation you could do:
```scala
queue.takeUpTo(bufferSize).flatMap(Fiber.collectAll(_)).map(ZChannel.writeChunk(_))
```

Reply:
Also the `SingleProducerAsyncInput` constructs seems to have been designed for `ZChannel.mergeAllWith`, but here there is only one consumer. It wouldn't be too hard to create a `SingleConsumerAsyncInput` that internalises the queue, that would drastically simplify your implementation. One element at a time processing instead of by chunks is the top reason for poor performance with streams (here is a nice example: https://blog.pierre-ricadat.com/supporting-high-performance-response-streaming-in-shardcake#heading-chunking-express)